keps/969-GangScheduling/README.md

```diff
@@ -0,0 +1,589 @@
+scheduled together or not at all. Partial scheduling wastes expensive GPU resources and blocks other jobs.
+
+With this KEP, I can specify:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: pytorch-training
+spec:
+  gangPolicy:
+    gangPolicyOption: JobSetAsGang
+  replicatedJobs:
+  - name: master
+    replicas: 1
+    template:
+      spec:
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: pytorch
+              image: pytorch/pytorch:latest
+              resources:
+                limits:
+                  nvidia.com/gpu: 8
+  - name: worker
+    replicas: 7
+    template:
+      spec:
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: pytorch
+              image: pytorch/pytorch:latest
+              resources:
+                limits:
+                  nvidia.com/gpu: 8
+```
+
+The scheduler will only schedule this JobSet when resources are available for all 64 GPUs simultaneously.
+
+#### Story 2: Per-ReplicatedJob Gang Scheduling
+
+As a HPC user running a multi-tier application with separate parameter servers and workers, I want each tier
+to be gang-scheduled independently. Parameter servers can start as soon as their gang is ready, while workers
+wait for their separate gang to be satisfied.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: multi-tier-training
+spec:
+  gangPolicy:
+    gangPolicyOption: JobSetGangPerReplicatedJob
+  replicatedJobs:
+  - name: ps
+    replicas: 4
+    template:
+      spec:
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: ps
+              image: tensorflow/tensorflow:latest
+  - name: worker
+    replicas: 16
+    template:
+      spec:
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: tensorflow/tensorflow:latest
+```
+
+Parameter server pods (4) will be gang-scheduled together, and worker pods (16) will be gang-scheduled as a separate group.
+
+#### Story 3: Advanced Workload Template Control
+
+As a ML engineer running distributed training, I have an initialization phase that must complete before training starts,
+but I want the leader and worker pods to be gang-scheduled together as a single unit (not separately). The simple gang
+policies don't support this pattern where I need custom grouping across ReplicatedJobs.
+
+With the WorkloadTemplate option, I can define custom pod sets that group the leader and worker together while keeping
+the initializer separate:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: multi-phase-training
+spec:
+  gangPolicy:
```

Comment:
Given that in the future Workload API will be integrated with TAS and DRA, how are we planning to integrate it into JobSet API?
- https://github.com/kubernetes/enhancements/pull/5711
- https://github.com/kubernetes/enhancements/pull/5733

If we plan to support templates via JobSet API, do we really need to introduce gangPolicy API?

One of the idea would be to directly go with `spec.workloadTemplate` API similarly to `VolumeClaimTemplates` in StatefulSet.

With that, we can gradually support more Workload features in JobSet controller.


cc @macsko  @wojtek-t @erictune @44past4 @johnbelamaric

Reply:
That was sorta my thought on Workload Template for more advanced features.

Same for supporting DependsOn with GangScheduling.